web/auth/interceptors/request_validator.go

```diff
@@ -0,0 +1,64 @@
+package interceptors
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/quickfeed/quickfeed/qf"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type validator interface {
+	IsValid() bool
+}
+
+type idCleaner interface {
+	RemoveRemoteID()
+}
+
+// RequestValidator returns a new unary server interceptor that validates requests
+// that implements the validator interface.
+// Invalid requests are rejected without logging and before it reaches any
+// user-level code and returns an illegal argument to the client.
+// In addition, the interceptor also implements a cancel mechanism.
+func RequestValidator(logger *zap.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		methodName := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
```

Comment:
Not sure we discussed it before, but the prometheus metrics stuff seems like a good target for a separate interceptor in `metrics.go`. Also, the `qf.Ag*` types should probably be moved to the interceptor package instead of the `qf` package? 

Similarly, I wonder if the following:
```go
	reg.MustRegister(
		promgrpc.NewServerMetrics(),
		qf.AgFailedMethodsMetric,
		qf.AgMethodSuccessRateMetric,
		qf.AgResponseTimeByMethodsMetric,
	)
```
could be move to the interceptor package's `init()` func. You would need to add a func like this:
```go
var reg = prometheus.NewRegistry()

func NewMetricsRegistry() *prometheus.Registry {
  return reg
}
```
So that you can get the `reg` to be passed to the http server. Or you could move this too:
```go
// Create an HTTP server for prometheus.
	httpServer := &http.Server{
		Handler: promhttp.HandlerFor(reg, promhttp.HandlerOpts{}),
		Addr:    fmt.Sprintf("127.0.0.1:%d", 9097),
	}
	go func() {
		if err := httpServer.ListenAndServe(); err != nil {
			log.Fatalf("Failed to start a http server: %v", err)
		}
	}()
```
And provide some other way to start it from main... but be aware of the `log.Fatalf` not being in `main.go` is an issue...

Reply:
I split the interceptors in this branch [grpc-interceptors](https://github.com/quickfeed/quickfeed/tree/grpc-interceptors).

Metrics: [metrics.go](https://github.com/quickfeed/quickfeed/blob/grpc-interceptors/qf/metrics.go)
Validation: [validation.go](https://github.com/quickfeed/quickfeed/blob/grpc-interceptors/qf/validation.go)
User Verifier: [auth.go](https://github.com/quickfeed/quickfeed/blob/83054cfbacd03b341db6abc85098f9703db507b4/web/auth/auth.go#L510)

Usage in `main.go`: [link](https://github.com/quickfeed/quickfeed/blob/83054cfbacd03b341db6abc85098f9703db507b4/main.go#L107)

`auth.go` contains both unary and streaming variants of the interceptor.

Reply:
Ok, should I then wait until it is ready and merged into master (and work on something else meanwhile)?